test/test/DottyDocTests.scala

```diff
@@ -0,0 +1,506 @@
+package test
+
+import dotty.tools.dotc.Compiler
+import dotty.tools.dotc.core.Phases.Phase
+import dotty.tools.dotc.typer.FrontEnd
+import dotty.tools.dotc.core.Contexts.Context
+import dotty.tools.dotc.ast.Trees._
+
+/** All tests need to extend this trait and define `source` and `assertion`
+ *
+ * Instances also need to be placed in the `DottyDocTests::tests` sequence to
+ * be included in the run
+ */
+trait DottyDocTest extends DottyTest { self =>
+  /** Source code in string format */
+  def source: String
+
+  /** A test to run against the resulting code */
+  def assertion: PartialFunction[Tree[Untyped], Unit]
+
+  private def defaultAssertion: PartialFunction[Tree[Untyped], Unit] = {
+    case x => assert(false, "Couldn't match resulting AST to expected AST in: " + x.show)
+  }
+
+  private def compiler(assertion: PartialFunction[Tree[Untyped], Unit]) = new Compiler {
+    override def phases = {
+      val checker = new Phase {
+        def phaseName = "assertionChecker"
+        override def run(implicit ctx: Context): Unit =
+          (assertion orElse defaultAssertion)(ctx.compilationUnit.untpdTree)
+      }
+
+      List(List(new FrontEnd)) ::: List(List(checker))
+    }
+  }
+
+  def checkDocString(actual: Option[String], expected: String): Unit = actual match {
+    case Some(str) => {
+      assert(str == expected, s"""Docstring: "$str" didn't match expected "$expected"""")
+    }
+    case None =>
+      assert(false, s"""No docstring found, expected: "$expected"""")
+  }
+
+  def run(): Unit = try {
+    val c = compiler(assertion)
+    c.rootContext(ctx)
+    c.newRun.compile(source)
+    println(s"${self.getClass.getSimpleName.split("\\$").last} passed")
+  } catch { case e: Throwable =>
+    Console.err.println(s"${self.getClass.getSimpleName.split("\\$").last} failed")
+    throw e
+  }
+}
+
+/** Add tests to the `tests` sequence */
+object DottyDocTests extends DottyTest {
+  private[this] val tests = Seq(
+    NoComment,
+
+    SingleClassInPackage,
+    MultipleOpenedOnSingleClassInPackage,
+    MultipleClassesInPackage,
+    SingleCaseClassWithoutPackage,
+    SingleTraitWihoutPackage,
+    MultipleTraitsWithoutPackage,
+    MultipleMixedEntitiesWithPackage,
+    NestedClass,
+    NestedClassThenOuter,
+    Objects,
+    ObjectsNestedClass,
+    PackageObject,
+    MultipleDocStringsBeforeEntity,
+    MultipleDocStringsBeforeAndAfter,
+
+    ValuesWithDocString,
+    VarsWithDocString,
+    DefsWithDocString,
+    TypesWithDocString
+  )
+
+  def main(args: Array[String]): Unit = {
```

Comment:
IE, you can make a method-per-every DottyDocTest, that only calls 

`checkCompile("frontend", source)(assertion)`